src/lib/mdns/ServicePool.cpp

```diff
@@ -0,0 +1,234 @@
+    }
+}
+
+CHIP_ERROR CopyTextEntries(MdnsService * dest, const MdnsService & source)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+    if (source.mTextEntrySize)
+    {
+        dest->mTextEntrySize = source.mTextEntrySize;
+        dest->mTextEntryies  = static_cast<TextEntry *>(chip::Platform::MemoryCalloc(source.mTextEntrySize, sizeof(TextEntry)));
+        VerifyOrExit(dest->mTextEntryies != nullptr, error = CHIP_ERROR_NO_MEMORY);
+
+        for (size_t i = 0; i < source.mTextEntrySize; i++)
+        {
+            TextEntry & entry = dest->mTextEntryies[i];
+
+            entry.mKey = static_cast<char *>(chip::Platform::MemoryAlloc(strnlen(source.mTextEntryies[i].mKey, kMdnsTextMaxSize)));
+            VerifyOrExit(entry.mKey != nullptr, error = CHIP_ERROR_NO_MEMORY);
+            entry.mDataSize = source.mTextEntryies[i].mDataSize;
+            entry.mData     = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(entry.mDataSize));
+            VerifyOrExit(entry.mData != nullptr, error = CHIP_ERROR_NO_MEMORY);
+        }
+    }
+
+exit:
+    if (error != CHIP_NO_ERROR)
+    {
+        FreeTextEntries(dest);
+    }
+
+    return error;
+}
+} // namespace
+
+namespace chip {
+namespace Mdns {
+
+ServicePool::ServicePool()
+{
+    ClearNoFree();
+}
+
+CHIP_ERROR ServicePool::AddService(uint64_t nodeId, uint64_t fabricId, const MdnsService & service, bool deepCopy)
+{
+    size_t hashValue = nodeId % kServicePoolCapacity;
+    size_t i         = hashValue;
+    bool found       = false;
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+    do
+    {
+        if (mEntries[i].mLazyDelete || mEntries[i].mNodeId == kUndefinedNodeId)
+        {
+            found = true;
+            break;
+        }
+
+        i++;
+        i %= kServicePoolCapacity;
+    } while (i != hashValue);
+
+    if (!found)
+    {
+        // Remove the entry that just matches the hash
+        i = hashValue;
+    }
+
+    if (mEntries[i].mLazyDelete)
+    {
+        mLazyDeleteCount--;
+        mEntries[i].mLazyDelete = false;
+    }
+    mEntries[i].mService  = service;
+    mEntries[i].mNodeId   = nodeId;
+    mEntries[i].mFabricId = fabricId;
+    if (deepCopy)
+    {
+        error = CopyTextEntries(&mEntries[i].mService, service);
+        if (error != CHIP_NO_ERROR)
+        {
+            mEntries[i].mNodeId = kUndefinedNodeId;
+        }
+    }
+    return error;
+}
+
+CHIP_ERROR ServicePool::RemoveService(uint64_t nodeId, uint64_t fabricId)
+{
+    size_t hashValue = nodeId % kServicePoolCapacity;
+    size_t i         = hashValue;
+
+    if (nodeId == kUndefinedNodeId)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    do
+    {
+        if (!mEntries[i].mLazyDelete && mEntries[i].mNodeId == kUndefinedNodeId)
+        {
+            break;
+        }
+        if (!mEntries[i].mLazyDelete && mEntries[i].mNodeId == nodeId && mEntries[i].mFabricId == fabricId)
+        {
+            mEntries[i].mLazyDelete = true;
+            mLazyDeleteCount++;
```

Comment:
Should there be a `ShouldRehash()` check at the end of this function followed by `Rehash()`?  Or is this the consumer's responsibility?

Reply:
Added rehash logic to `DiscoveryManager`. We want the consumer to control the rehash so that the latency becomes more predictable.